source/blog/_posts/2016-09-05-psh-managing-build-complexity.md

```diff
@@ -0,0 +1,290 @@
+
+* `properties.yml` - as Symfony configuration
+* `docker-compose.override.yml` - Docker settings in a Linux-DEV and in a CI version
+* `Vagrantfile` - Vagrant settings for Mac-Developers
+* `build.properties` - The ANT properties
+
+A few of these files have shared values, e.g the database server settings were shared between the application, the build system and the virtual machine hosting the database. 
+But in reality all these tools use different formats to basically share the same values. ANT as the center of our build System has no easy way of propagating these values.
+Stranded with the alternative of writing one string replace per value, or copying the whole file over and over we were helpless in reducing the lines of code necessary for our environment.
+A templating engine would have solved this easily and naturally.
+
+##### Platform independence
+
+Although ANT-Scripts can be executed on multiple platforms hardly any script can. If you a) are not developing a Java application and b) need a little more then simple file copies
+you will sooner or later start translating [shell commands to ANT XML](https://ant.apache.org/manual/Tasks/exec.html), and let it execute them. For us this happened 
+with `composer`, `phpunit`, `nodejs`, `ansible` and many other cases. By that your application is limited by the availability of the shell. So why write XML around it?
+
+##### User Experience
+
+ANT simply is no fun to use. Most failures in script development were not because the thing I tried to achieve were hard to do,
+but ANT stands in the way. Either I misread the 20 lines of XML, I need for a simple statement, or I anticipated different [properties](https://ant.apache.org/manual/Tasks/property.html)
+to be present. My solution for this was to use [macros](/blog/2016/03/09/managing-ant-complexity#macros) extensively. This certainly helps a lot, but I must say I would prefer a tool that
+would not give me an existential crisis every time I had to change something about it.
+
+### Starting a new... 
+ 
+This all meant for us that all XML based build tools and especially ANT were no longer an option. While looking around for a ANT alternative we were starting our development environment
+by sharing a few sh scripts and using them to ensure that we could all work - it simply is the most natural format, you type a statement, execute it, verify it worked and copy and 
+paste it into a `script.sh` file. Everybody can do it and so it no longer was solely my task to create the environment we would all work in. Times were great! 
+
+But over the process of a one or two weeks some problems started to arise. We had gained a `*.dist` file, Docker had a different MySQL host then local development which meant copying whole 
+scripts just to change one or two lines. And worst of all: **If a statement at the beginning of shell script fails execution continues** - something that you will notice far too late.
```

Comment:
You should probably explain why this problem could not be solved by adding `set +e` on top of the bash scripts.